mobius-core/src/test/java/com/spotify/mobius/MobiusLoopTest.java

```diff
@@ -440,20 +446,144 @@ public void eventsFromEffectHandlerDuringDisposeAreIgnored() throws Exception {
     assertFalse(updateWasCalled.get());
   }
 
+  @Test
+  public void disposingLoopWhileInitIsRunningDoesNotEmitNewState() throws Exception {
+    // Model changes emitted from the init function during dispose should be ignored.
+
+    // This test sets up the following scenario:
+    // 1. The loop is created and initialized but the Init function is blocked
+    // 2. The loop is created with an EventSource that returns a Disposable that blocks whenever invoked
+    // this is so we can block the loop's dispose call and unblock it on demand
+    // 3. Starts a new thread that attempts to dispose of the loop
+    //
+    // Once loop.dispose() is invoked, the loop will attempt to dispose of its EventSource disposable
+    // which will block. At this point, both Init and loop.dispose() are blocked, both on the same
+    // semaphore (i.e. two permits have been requested from the semaphore).
+    // We release 2 permits for the semaphore which would unblock both Init and loop.dispose().
+    // We then request 2 permits to block the test until both dispose and Init have finished.
+    // We release the first permit after loop.dispose() returns, and the second after the
+    // initialization runnable runs. By now, both initialization and disposal have been completed
+    // and we can proceed to inspect our observer.
+
+    observer = new RecordingModelObserver<>();
+    Semaphore initLock = new Semaphore(0);
+    Semaphore awaitDisposalRequest = new Semaphore(0);
+    Semaphore disposalCompleted = new Semaphore(0);
+
+    final Update<String, TestEvent, TestEffect> update = (model, event) -> Next.noChange();
+    final MobiusLoop.Builder<String, TestEvent, TestEffect> builder =
+        Mobius.loop(update, effectHandler)
+            .init(
+                m -> {
+                  initLock.acquireUninterruptibly();
+                  return First.first(m);
+                })
+            .eventSource(c -> initLock::acquireUninterruptibly)
```

Comment:
@pettermahlen  I had both init and dispose block on the same semaphore, then I released two permits so they'd both unblock simultaneously. This should do what you were referring to.

Reply:
If you really want to guarantee that `init` is running when `dispose` gets called, you would need `init` to allow the test code to proceed to invoking dispose once it's running. You should be able to reuse the `awaitDisposalRequest` semaphore for that, and wait for 2 releases before invoking `dispose`.